api/src/nv_ingest_api/internal/primitives/nim/nim_client.py

```diff
@@ -390,6 +456,137 @@ def _http_infer(self, formatted_input: dict) -> dict:
         logger.error(f"Failed to get a successful response after {self.max_retries} retries.")
         raise Exception(f"Failed to get a successful response after {self.max_retries} retries.")
 
+    def _batcher_loop(self):
+        """The main loop for the background thread to form and process batches."""
+        while not self._stop_event.is_set():
+            requests_batch = []
+            try:
+                first_req = self._request_queue.get(timeout=self._batch_timeout)
+                if first_req is None:
+                    continue
+                requests_batch.append(first_req)
+
+                current_max_h = first_req.dims[0]
+                current_max_w = first_req.dims[1]
+
+                start_time = time.monotonic()
+
+                while len(requests_batch) < self._batch_size:
+                    if (time.monotonic() - start_time) >= self._batch_timeout:
+                        break
+
+                    if self._batch_memory_budget_bytes:
+                        try:
+                            next_req_peek = self._request_queue.queue[0]
+
+                            next_h = max(current_max_h, next_req_peek.dims[0])
+                            next_w = max(current_max_w, next_req_peek.dims[1])
+                            potential_batch_size = len(requests_batch) + 1
+
+                            num_channels = 3
+                            bytes_per_element_fp32 = 4
+                            potential_memory_bytes = (
+                                potential_batch_size * next_h * next_w * num_channels * bytes_per_element_fp32
+                            )
```

Comment:
This logic is probably dependent on model/values we are sending and may change for other models. Lets abstract/send this logic to a model specific code if possible